nova/virt/vmwareapi/driver.py

```diff
@@ -690,6 +697,250 @@ def detach_interface(self, context, instance, vif):
         """Detach an interface from the instance."""
         self._vmops.detach_interface(context, instance, vif)
 
+    @periodic_task.periodic_task(spacing=CONF.vmware.drs_group_sync_interval)
+    def _sync_server_groups(self, compute_manager, context):
+        """Regularly syncs the DB state of server-groups with the cluster
+
+        This function retrieves all groups/rules from the cluster, all
+        instances belonging to this host and all server-groups belonging to
+        these instances. After building a list of expected groups/rules and the
+        expected members from those server-groups, it compares them to the
+        retrieved DRS groups/rules building a list of changes to bring the
+        cluster into expected state. At the end, the changes are applied in
+        batches.
+
+        We only sync DRS groups/rules if they have more then 2 members. We
+        would waste resources on the vCenter otherwise.
+
+        This function is different from other periodic_task's in that it
+        receives the manager as first argument. This comes from the fact, that
+        it's not part of that manager, but instead only registered to it, so it
+        can be called. The argument is unused in this function.
+        """
+        LOG.debug('Starting DRS groups/rules sync-loop.')
+        client_factory = self._session.vim.client.factory
+
+        # retrieve all DRS groups, rules and members
+        cluster_groups = cluster_util.fetch_cluster_groups(
+            self._session, cluster_ref=self._cluster_ref, group_type='vm')
+        cluster_rules = cluster_util.fetch_cluster_rules(
+            self._session, cluster_ref=self._cluster_ref)
+
+        # NOTE: No need to filter the members of DRS groups and rules, because
+        # DRS rules/groups are cluster-local and thus can only contain VMs of
+        # the cluster.
+
+        # filter DRS groups matching to our naming scheme
+        cluster_groups = {name: g for (name, g) in cluster_groups.items()
+                          if name.startswith(constants.DRS_PREFIX)}
+        cluster_rules = {name: r for (name, r) in cluster_rules.items()
+                         if name.startswith(constants.DRS_PREFIX)}
```

Comment:
I disagree. Operators might change rules. External scripts might changes rules (e.g. for the live-migration case, but there might be other stuff). I wouldn't want to work on an outdated cache.

That being said, we also have a vmref cache, that can very easily get outdated via operator actions, so maybe it's worth the risk. But I wouldn't want to rely on a cache, without a way to recover from errors, that could be caused by the cache containing an outdated rule - thus added complexity in the sync function.

Reply:
I think, cache is the wrong term here, and the use-cases you listed are actually a reason to work with a local copy: To properly identify the intend.

Yes, we need to be able to ensure that the model in nova is correctly reflected in the vcenter, but if the operator changing rules and external scripts changing the rules are valid use-cases, we need that to happen rather rarely and at well-defined times. Preferably controllable by whoever does the manual steps outside of nova.

If you base the decision on what is stored in the vcenter, you will change the rules if there is a discrepancy to what you expect. So, if the live-migration script is started, and removes the VM from a rule-set so it is actually possible to move the VM, and then the sync-task comes along and discovers that, and reapplies the rule, then it will break the script. Similar to whatever an Operator might want to do manually.

So, in that case, the sync task needs to run rarely (a couple of times per day?) to give operators / scripts some time-window to do what they need to do.